File: Python/014.py
import time
start = time.time()
def collatz_length(n):
    count = 0
    sequence = []
    while True:
        if n in table:
            length = table[n] + count
            for i in range(len(sequence)):
                table[sequence[i]] = length-i
            return length
        else:
            sequence.append(n)
            if n%2 == 0:
                n = n//2
            else:
                n = 3*n+1
            count += 1
            
            
table = {1:1}
max = 0
for i in range(1,1000000):
    length = collatz_length(i)
    if length > max:
        max = length
        index = i
print(index)
print(time.time()-start)

File: C++/014.cpp
#include <iostream>

using namespace std;

int collatzLength(long long int n);

int collatzLength(long long int n) {
    int count = 0;
    while (1) {
      if (n == 1)
        return count + 1;
      if (n%2 == 0)
        n /= 2;
      else
        n = 3*n + 1;
      count++;
    }
}

int main() {
  int max = 0, index = 0;
  for (int i=1; i<1000000; i++) {
    int length = collatzLength(i);
    if (length > max) {
      max = length;
      index = i;
    }
  }
  cout << index << endl;
}
